### ISSUE TITLE
Code has higher MI w/ Concept in Pretraining

### ISSUE THREAD

**Issue by TerryTong-Git** [TerryTong-Git/ToolProj#20]
## Description

Since we found that code has higher MI w/ concept during inference, we want to try motivate this. One idea is that this may stem from pretraining. 

## Hypothesis or Goal

On pretraining documents, code generally helps discriminate the tasks much better. This corresponds to a lower cross entropy loss for the code over the natural language, when trained to predict the concept labels.  

### Methodology 

1. I will filter the datasets and cluster them heuristically based off of the words that are present and the numbers that are present. Heuristics are scalable so I use them
2. I will use these heuristic numbers as gold labels and train a logistic regression to try predict the model. Will train a model for NL and for code, controlling the number of samples, average length of samples, and num samples per label. Then will try see the cross entropy I get. 
3. Report the cross entropy as well as the discrimination accuracy. 

### Experimental Setup
I'll try heuristic word filtering, or llama embedding features filtering. I suspect the first one is much faster, so I'll try that first and see the quality. 

For example: 
```
CONCEPT_KEYWORDS = {
    "knapsack_dp": ["knapsack", "0/1", "maximize value", "weight limit"],
    "ilp": ["integer linear program", "ilp", "lp relaxation", "constraints"],
    "dp": ["dynamic programming", "subproblem", "recurrence"],
    "arithmetic": ["add", "subtract", "multiply"],
}. 
```
Then for each of these subsets, I'll filter based on the numbers that are present, and bin them, like how I did during the inference procedure. 

For NL I'll filter [Wikipedia](https://huggingface.co/datasets/wikimedia/wikipedia), and for code I'll filter [theStackV2](bigcode/the-stack-v2) and compare. For each of these, I'll collect some dataset statistics, such as the number of tokens under a GPT2 tokenizer, the number of datapoints that I'm keeping, samples per label etc. just to make the experiment fair. Otherwise, I'll filter via the BERT or LLama3.1 embeddings. 

Each cluster will correspond to a latent concept, then I'll label these with that latent concept. I'll manually verify that these makes sense by randomly sample like 10 from each cluster and check them. 

### Estimated / Success Probability

Since this worked for the MI in inference, I am pretty sure it will work for the pretraining data too, so I guess around 80% success rate. I won't take long since I already have the MI infrastructure to compute it, I just need to download, subsample, filter, and cluster the pretraining data, then train LR, which may take a while. 

### Links

* WandB Report:  (link)

## Results

TODO

**Comment by TerryTong-Git** [comment]
Q1:
I decided to embed the chunks and try using retrieval to get code / NL closest to the algorithmic concept, and inspect a few of the retrieved to make sure. I'm using miniLM as the embedding / retrieval model. I plan to use the queries I defined above and see how it does. I didn't use LLM-as-a-judge because it can't scale for pretraining datasets of sizes > 1B datapoints. 

Q2:
For the datasets, I switched to StarCoder and DCLM-baseline since the data seemed cleaner and already chunked. These larger pretraining datasets have other pretraining datasets as a subset. For context, starcoder is 450Billion tokens, and DCLM baseline has 4 Trillion tokens. 

Q3:
I am running on Marin-8B-instruct, will see what happens.

**Comment by AnnieYuFeng** [comment]
Using explicit algorithm (query) terms to seed MiniLM retrieval may still bias results toward code that literally names the concept, making code look more discriminative than it really is. You could try descriptor paraphrases of the algorithm.

I also have concerns about whether we can just use this experiment to explain the phenomenon, as there are many other confounds in pretraining. Even if we show that, on average, code data better predicts the target concepts, we have no control over how the model learns from each type of data during pretraining, or over the relative size of the code versus natural language data. So other than the data's different predicting the concept abilities(which could already be biased due to how you choose the gold label for the logistic regresstion), code and natural language differ in length, tokenization, and structure, which may lead to different learning efficiency of the algorithm in pretraining, and pretraining data exposure(the size of the codes in the pretraining data about the algorithms vs in natural language) may also play a role here. 

I would still feel that we need a more careful design of the pretraining experiment.
